addons/payment_ingenico/models/payment.py

```diff
@@ -29,13 +30,27 @@ class PaymentAcquirerOgone(models.Model):
     ogone_pspid = fields.Char('PSPID', required_if_provider='ogone', groups='base.group_user')
     ogone_userid = fields.Char('API User ID', required_if_provider='ogone', groups='base.group_user')
     ogone_password = fields.Char('API User Password', required_if_provider='ogone', groups='base.group_user')
-    ogone_shakey_in = fields.Char('SHA Key IN', size=32, required_if_provider='ogone', groups='base.group_user')
```

Comment:
> I believe the size attribute is passed to PostgreSQL to allocate the requested memory (so we should be fine) for the table.

No, it's used to set a constraint on the table / column. Postgres always uses flexible string storage, IIRC it doesn't even store fixed-size CHARs differently, they just get automatically padded on insertion.

Reply:
oh right I missed that, indeed the size is set at the SQL level so you will get an issue...
```
 ogone_shakey_in          | character varying(32)       |           |          | 
 ogone_shakey_out         | character varying(32)       |           |          | 
```
```
psycopg2.errors.StringDataRightTruncation: value too long for type character varying(32)
```

Reply:
I based my solution on what I read in https://github.com/odoo/odoo/wiki/Contributing#what-does-stable-mean
> Limited, compatible changes such as changing ondelete rules or size parameters are allowed when necessary.
+ Testing creating the DB in 13.0 and then adding the change and there was no error (nor on setting 64-long char value in the fields)

But if size change are not possible nor is adding new field, what do you suggest to save the new fields?

@AntoineVDV What do you mean by `key is chosen by the user` do you mean the user can choose the type of encoding of the key ?
Or ingenico allow for users to say : my key must be 'mdp'  and then they just pad to correspond to SHA specs ?

Reply:
On the other hand, dropping the fixed size won't break any *currently working* setup, so it's not that bad. But it won't get older databases to a fully working state, as SHA256/512 keys will get truncated upon saving.

You might be able to detect that case by overriding `write()`, and when the length of the `ogone_shakey*` is > 32, you try to `flush()` it and re-read it immediately. If the value is truncated after reading it, the `payment_ingenico` module needs to be updated manually, and you can raise an error to let the user know.

It's a bit ugly to add short-lived ad hoc error handling like that, but it may only take a few lines of code, and can be dropped immediately in `saas-15.2`/`master`. If the fix is really necessary to support ogone payments correctly, it may be worth it.

Reply:
It seems to me that we don't need to take such measures because the user can choose the value of both the SHA-IN and SHA-OUT keys (see screenshot). This is what I meant @fdamhaut. 

![image](https://user-images.githubusercontent.com/15896813/156757645-e72a351a-8576-4ced-83d2-70bfa762be41.png)

If Odoo limits the keys' size to 32 in 13.0, no big deal: the user truncates their keys so that they are accepted by Odoo.

What we need to know is whether the signature was computed with SHA1/256/512. Given that the key length is arbitrary, I see no other way than asking the user which algorithm is configured on their Ingenico back office. In stable, this can be achieved through an optional `ir.config.parameter` while, in master, it should be a new `Selection` field hidden in debug mode and with 'SHA512' as default value and 'SHA1' as "migrated" value.

Reply:
> It seems to me that we don't need to take such measures because the user can choose the value of both the SHA-IN and SHA-OUT keys (see screenshot).

Oh that's great!

> If Odoo limits the keys' size to 32 in 13.0, no big deal: the user truncates their keys so that they are accepted by Odoo (this is what I meant @fdamhaut).

Well, the user won't actually know that they have to do that, as the value would be silently truncated upon being saved.

But indeed, if we have to write some documentation about the special ICP that is needed in stable to select the hashing algorithm, we can let users know about the necessary module update as well :+1: